lib/datadog/grok/src/parse_grok.rs

```diff
@@ -0,0 +1,289 @@
+    grok_patterns
+        .iter()
+        .filter(|&rule| {
+            rule.destination.is_some() && rule.destination.as_ref().unwrap().filter_fn.is_some()
+        })
+        .for_each(|rule| {
+            let dest = rule.destination.as_ref().unwrap();
+            filters
+                .entry(dest.path.clone())
+                .and_modify(|v| v.push(dest.filter_fn.as_ref().unwrap().clone()))
+                .or_insert(vec![dest.filter_fn.as_ref().unwrap().clone()]);
+        });
+
+    let mut pattern = String::new();
+    pattern.push('^');
+    pattern.push_str(rule_def.as_str());
+    pattern.push('$');
+
+    // store rule definitions and filters in case this rule is referenced in the next rules
+    prev_rule_patterns.insert(rule_name, rule_def);
+    prev_rule_filters.insert(rule_name, filters.clone());
+
+    Ok(GrokRule { pattern, filters })
+}
+
+/// Converts each rule to a pure grok rule:
+///  - strips filters and "remembers" them to apply later
+///  - replaces references to previous rules with actual definitions
+///  - replaces match functions with corresponding regex groups
+fn purify_grok_pattern(
+    prev_rule_definitions: &mut HashMap<&str, String>,
+    prev_rule_filters: &mut HashMap<&str, HashMap<LookupBuf, Vec<Function>>>,
+    mut filters: &mut HashMap<LookupBuf, Vec<Function>>,
+    rule: &GrokPattern,
+) -> Result<String, Error> {
+    let mut res = String::new();
+    if prev_rule_definitions.contains_key(rule.match_fn.name.as_str()) {
+        // this is a reference to a previous rule - replace it and copy all destinations from the prev rule
+        res.push_str(
+            prev_rule_definitions
+                .get(rule.match_fn.name.as_str())
+                .unwrap(),
+        );
+        if let Some(d) = prev_rule_filters.get(rule.match_fn.name.as_str()) {
+            d.iter().for_each(|(path, function)| {
+                filters.insert(path.to_owned(), function.to_owned());
+            });
+        }
+        Ok(res)
+    } else if rule.match_fn.name == "regex"
+        || rule.match_fn.name == "date"
+        || rule.match_fn.name == "boolean"
+    {
+        // these patterns will be converted to named capture groups e.g. (?<http.status_code>[0-9]{3})
+        res.push_str("(?");
+        res.push_str("<");
+        if let Some(destination) = &rule.destination {
+            res.push_str(destination.path.to_string().as_str());
+        }
+        res.push_str(">");
+        res.push_str(process_match_function(&mut filters, &rule)?.as_str());
+        res.push_str(")");
+
+        Ok(res)
+    } else {
+        // these will be converted to "pure" grok patterns %{PATTERN:DESTINATION} but without filters
+        res.push_str("%{");
+
+        res.push_str(process_match_function(&mut filters, &rule)?.as_str());
+
+        if let Some(destination) = &rule.destination {
+            res.push_str(":");
+            res.push_str(destination.path.to_string().as_str());
```

Comment:
```suggestion
            write!(res, ":{}", destination.path).unwrap();
```

Note: This requires `std::fmt::Write` in scope.

Reply:
Done 👍